### PR TITLE
Add callback to terminate training if NaN loss encountered. (Update to #4849)

### PR DESCRIPTION
**Pull request by cengh**
In cases where jobs are submitted to a queue, it would be helpful to have training terminated automatically when NaN loss is encountered, as stated in issue #4848. This PR uses a callback to stop training at the end of a batch if the loss is NaN.

### CODE REVIEW

File: keras/callbacks.py [hunk 1]
```diff
@@ -227,6 +228,33 @@ def on_epoch_end(self, epoch, logs=None):
                     logs[k] = self.totals[k] / self.seen
 
 
+class TerminateOnNaN(Callback):
+    '''Callback that terminates training when a nan loss is encountered.
+
+    # Arguments
+        patience: number of mini-batches with invalid loss
```
Comment (fchollet) [inline keras/callbacks.py]: Patience does not make sense in this context: NaN is unrecoverable.

File: keras/callbacks.py [hunk 2]
```diff
@@ -227,6 +228,33 @@ def on_epoch_end(self, epoch, logs=None):
                     logs[k] = self.totals[k] / self.seen
 
 
+class TerminateOnNaN(Callback):
+    '''Callback that terminates training when a nan loss is encountered.
```
Comment (fchollet) [inline keras/callbacks.py]: `NaN`

File: keras/callbacks.py [hunk 3]
```diff
@@ -227,6 +228,33 @@ def on_epoch_end(self, epoch, logs=None):
...
+        current = logs.get('loss')
+        if np.isnan(current) or np.isinf(current):
+            self.invalid += 1
+        else:
+            self.invalid = 0
+
+        try:
+            assert self.invalid <= self.patience
+        except AssertionError:
+            print('Batch %05d: %d consecutive mini-batches with invalid loss, terminating training' % (batch, self.patience))
```
Comment (fchollet) [inline keras/callbacks.py]: Line too long

File: keras/callbacks.py [hunk 4]
```diff
@@ -227,6 +228,33 @@ def on_epoch_end(self, epoch, logs=None):
...
+    def on_batch_end(self, batch, logs={}):
+        # terminate training immediately if nan loss encountered in a batch
+        current = logs.get('loss')
+        if np.isnan(current) or np.isinf(current):
+            self.invalid += 1
+        else:
+            self.invalid = 0
+
+        try:
+            assert self.invalid <= self.patience
```
Comment (fchollet) [inline keras/callbacks.py]: `try` with an `assert` is not the way to test inequality between two integers...

File: keras/engine/training.py [hunk 5]
```diff
@@ -1124,52 +1124,57 @@ def _fit_loop(self, f, ins, out_labels=None, batch_size=32,
...
-                batch_logs['size'] = len(batch_ids)
-                callbacks.on_batch_begin(batch_index, batch_logs)
-                outs = f(ins_batch)
-                if not isinstance(outs, list):
-                    outs = [outs]
-                for l, o in zip(out_labels, outs):
-                    batch_logs[l] = o
-
-                callbacks.on_batch_end(batch_index, batch_logs)
+        try:
```
Comment (fchollet) [inline keras/engine/training.py]: No, do not do this

File: keras/engine/training.py [hunk 6]
```diff
@@ -1124,52 +1124,57 @@ def _fit_loop(self, f, ins, out_labels=None, batch_size=32,
...
+                                                       verbose=0)
+                            if not isinstance(val_outs, list):
+                                val_outs = [val_outs]
+                            # Same labels assumed.
+                            for l, o in zip(out_labels, val_outs):
+                                epoch_logs['val_' + l] = o
+                callbacks.on_epoch_end(epoch, epoch_logs)
+                if callback_model.stop_training:
+                    raise Exception
+        except Exception:
```
Comment (fchollet) [inline keras/engine/training.py]: wtf

File: tests/keras/test_callbacks.py [hunk 7]
```diff
@@ -23,6 +24,35 @@
...
+    y_test = np_utils.to_categorical(y_test)
+    y_train = np_utils.to_categorical(y_train)
+    cbks = [callbacks.TerminateOnNaN()]
+    model = Sequential()
+    initializer = initializers.Constant(value=1e5)
+    for _ in range(5):
+        model.add(Dense(num_hidden, input_dim=input_dim, activation='relu',
+                        kernel_initializer=initializer))
+    model.add(Dense(num_class, activation='linear'))
+    sgd = optimizers.SGD(lr=np.nan)
```
Comment (Dref360) [inline tests/keras/test_callbacks.py]: Not used?

Comment (cengh) [inline tests/keras/test_callbacks.py]: Thanks. Line has been removed.

File: keras/callbacks.py [hunk 8]
```diff
@@ -227,6 +227,18 @@ def on_epoch_end(self, epoch, logs=None):
                     logs[k] = self.totals[k] / self.seen
 
 
+class TerminateOnNaN(Callback):
+    """Callback that terminates training when a NaN loss is encountered."""
+    def __init__(self):
+        super(TerminateOnNaN, self).__init__()
+
+    def on_batch_end(self, batch, logs=None):
+        current = logs.get('loss')
```
Comment (fchollet) [inline keras/callbacks.py]: This should be:
```python
logs = logs or {}
loss = logs.get('loss')
if loss is not None:
   if np.isnan(loss) or np.isinf(loss):
```

File: keras/callbacks.py [hunk 9]
```diff
@@ -227,6 +227,18 @@ def on_epoch_end(self, epoch, logs=None):
...
+        if np.isnan(current) or np.isinf(current):
+            print('Batch %05d: Invalid loss, terminating training' % (batch))
```
Comment (fchollet) [inline keras/callbacks.py]: While print the batch index with so many leading zeros? Just use `%d`

File: keras/callbacks.py [hunk 10]
```diff
@@ -227,6 +227,18 @@ def on_epoch_end(self, epoch, logs=None):
                     logs[k] = self.totals[k] / self.seen
 
 
+class TerminateOnNaN(Callback):
+    """Callback that terminates training when a NaN loss is encountered."""
```
Comment (fchollet) [inline keras/callbacks.py]: Insert a line break after the docstring

Comment (cengh) [inline keras/callbacks.py]: Thanks for the feedback! The changes have been incorporated.

**Comment by fchollet** [comment]
LGTM, thanks

**Comment by vfdev-5** [comment]
I just wonder why this is not directly integrated into the training loop ? 
Are there cases where we would be ok to continue training with a nan/inf loss ?
Thanks
